server/src/test/java/org/elasticsearch/common/util/concurrent/AbstractAsyncTaskTests.java

```diff
@@ -0,0 +1,203 @@
+            }
+
+            @Override
+            protected void runInternal() {
+                assertTrue("generic threadpool is configured", Thread.currentThread().getName().contains("[generic]"));
+                final CountDownLatch l1 = latch1.get();
+                final CountDownLatch l2 = latch2.get();
+                count.incrementAndGet();
+                l1.countDown();
+                try {
+                    l2.await();
+                } catch (InterruptedException e) {
+                    fail("interrupted");
+                }
+                if (randomBoolean()) {
+                    throw new RuntimeException("foo");
+                }
+            }
+
+            @Override
+            protected String getThreadPool() {
+                return ThreadPool.Names.GENERIC;
+            }
+        };
+
+        assertFalse(task.isScheduled());
+        task.rescheduleIfNecessary();
+        assertTrue(task.isScheduled());
+        latch1.get().await();
+        latch1.set(new CountDownLatch(1));
+        assertEquals(1, count.get());
+        // here we need to swap first before we let it go otherwise threads might be very fast and run that task twice due to
+        // random exception and the schedule interval is 1ms
+        latch2.getAndSet(new CountDownLatch(1)).countDown();
+        latch1.get().await();
+        assertEquals(2, count.get());
+        assertTrue(task.isScheduled());
+        task.close();
+        assertTrue(task.isClosed());
+        assertFalse(task.isScheduled());
+        latch2.get().countDown();
+        assertEquals(2, count.get());
+    }
+
+    public void testManualRepeat() throws Exception {
+
+        final CyclicBarrier barrier = new CyclicBarrier(2); // 1 for runInternal plus 1 for the test sequence
+        final AtomicInteger count = new AtomicInteger();
+        AbstractAsyncTask task = new AbstractAsyncTask(logger, threadPool, TimeValue.timeValueMillis(1), false) {
+
+            @Override
+            protected boolean mustReschedule() {
+                return true;
+            }
+
+            @Override
+            protected void runInternal() {
+                assertTrue("generic threadpool is configured", Thread.currentThread().getName().contains("[generic]"));
+                count.incrementAndGet();
+                try {
+                    barrier.await(10, TimeUnit.SECONDS); // should happen very quickly
+                } catch (Exception e) {
+                    fail("barrier not passed in reasonable time");
+                }
+                if (randomBoolean()) {
+                    throw new RuntimeException("foo");
+                }
+            }
+
+            @Override
+            protected String getThreadPool() {
+                return ThreadPool.Names.GENERIC;
+            }
+        };
+
+        assertFalse(task.isScheduled());
+        task.rescheduleIfNecessary();
+        barrier.await(10, TimeUnit.SECONDS); // should happen very quickly
```

Comment:
we typically wait indefinitely so if things get stuck we get a suite timeout +  stack dump so we know where things got stuck